### PR TITLE
net/gcoap: Create confirm request infrastructure and adapt existing messaging

### PR DESCRIPTION
**Pull request by kb2ma**
This PR implements the second step in the confirmable messaging plan in #7192. It extends the request memo struct to support sending, and resending, a message of type CON. In addition, the existing message sending code has been adapted to use the changed struct.

This PR allows sending a request of CON type, but the message will not be resent if the response times out. The gcoap example also has been updated to specify use of the CON type, by addition of a `-c` option. Notice the call to `coap_hdr_set_type()` after initializing the PDU in `gcoap_cli_cmd()`.

The resend implementation and documentation on sending a confirmable request are the subjects of the next PR in the series. I have started to implement it as a WIP (#7337) to verify that the present PR provides what's required.

### CODE REVIEW

**Comment by haukepetersen** [comment]
Just been playing a little with this PR and I think I found an issue: I am sending a `CON` request (`POST` to `/.well-known/core` to do a simplified registration with a RD). The server responds with an `ACK`., that has an empty payload. This behavior is intended, but `gcoap` gets caught by `empty messages not handled yet`...

As I understood this PR, it should actually handle empty messages (or at least empty ACK messages by now), right?

**Comment by kb2ma** [comment]
An empty (code 0.00) ACK indicates the server intends to see the separate response later. See Sec. 2.2 of the spec. This scenario is not included in this PR because it would require the client to retain the request memo after the initial ACK response. Handling for a separate response is the sixth item in #7192.

**Comment by haukepetersen** [comment]
Yes, makes sense. Seems like I got something wrong when debugging, now I can't even reproduce the behavior I've seen on Monday... So never mind I guess.

**Comment by kb2ma** [comment]
Rebased.

**Comment by kb2ma** [comment]
Thanks, @haukepetersen. This PR is a stepping stone to the main goal of merging #7337. I plan to rebase that one in the next day or so.

**Comment by kb2ma** [comment]
Apologies, all. I dropped a commit in the rebase. Must redo it.

**Comment by gebart** [comment]
@kb2ma you can use the `git reflog` command to find your lost commit of you haven't manually pruned your local clone. Use git cherry-pick to reapply the commit.

**Comment by kb2ma** [comment]
Actually, the commits here are OK. I got confused between what belongs in this PR and what belongs in #7337.

@gebart -- thanks for the tip. Actually, my first rule of rebasing is to copy the local repository before starting. ;-)

**Comment by kaspar030** [comment]
> My first rule of rebasing is to copy the local repository before starting. ;-)

Using a temporary branch is usually much faster. ```git branch <whatever>_tmp; git checkout <whatever>_tmp; git rebase <other>;```, and if everything works out, ```git checkout <whatever>; git reset --hard <whatever_tmp>; git branch -D <whatever_tmp>```

**Comment by miri64** [comment]
> My first rule of rebasing is to copy the local repository before starting. ;-)

Why? As long as you don't `git gc --purge` on purpose git is storing everything you put into the index at some point. In case you mess up your rebase: `git reflog` (log of all previous `HEAD` commits) is your friend ;-)

File: sys/net/application_layer/gcoap/gcoap.c [hunk 1]
```diff
@@ -331,39 +330,41 @@ static ssize_t _finish_pdu(coap_pkt_t *pdu, uint8_t *buf, size_t len)
...
-            size_t j;
-            for (j = 0; j < coap_get_token_len(src_pdu); j++) {
-                if (*src_byte++ != *memo_byte++) {
-                    break;      /* token mismatch */
+        else {
+            memo_pdu->hdr = (coap_hdr_t *) memo->msg.data.pdu_buf;
+        }
+
+        if (coap_get_token_len(memo_pdu) == cmplen) {
+            if (cmplen) {
```
Comment (smlng) [inline sys/net/application_layer/gcoap/gcoap.c]: this can be optimised, i.e., by merging the 2 `if` to `if (cmplen && (coap_get_token_len(memo_pdu) == cmplen)) {`. However, I would rather put this into a helper function `_cmp_hdr_token(token A, token B)` which returns `-1`, `0`, or `1` like typical compare functions. That would make this a one line and this whole for-loop a lot more readable (IMHO).

Comment (kb2ma) [reply sys/net/application_layer/gcoap/gcoap.c]: I just looked deeper into `memcmp()`, and I was wrong about the need to explicitly test when the comparison length is zero, as explained in this [SO question](https://stackoverflow.com/questions/19034858/return-value-of-memcmpptr1-ptr2-0).

So, the token comparison can be simplified to:
```
        if (coap_get_token_len(memo_pdu) == cmplen) {
            memo_pdu->token = &memo_pdu->hdr->data[0];
            if (memcmp(src_pdu->token, memo_pdu->token, cmplen) == 0) {
                *memo_ptr = memo;
                break;
            }
        }
```

Good enough?

**Comment by smlng** [comment]
please squash

**Comment by kb2ma** [comment]
Squashed and rebased.

**Comment by smlng** [comment]
@haukepetersen can you spare some time and have a look too, I think we should move this and related PRs by @kb2ma forward, otherwise I'm also fine to push this on my own.
